### ISSUE TITLE
Error Creating a container with BCContainerHelper -- New-NavContainer.ps1:1828 Character:15

### ISSUE THREAD

**Issue by MarcFB** [microsoft/navcontainerhelper#2419]
Hello

I try to set up a container based development environment according to 
https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-running-container-development

I use the "BCContainerHelper" section:
- install the moule with "install-module BCContainerHelper -force"
- the using the command "$artifactUrl = Get-BcArtifactUrl -type sandbox -country us -select Latest
New-BCContainer -accept_eula -containerName mysandbox -artifactUrl $artifactUrl" 

everything as witten in the section.

When I execute the command, I get the following output

S C:\WINDOWS\system32> $artifactUrl = Get-BcArtifactUrl -type sandbox -country ch -select Latest
New-BCContainer -accept_eula -containerName mysandbox -artifactUrl $artifactUrl
BcContainerHelper is version 3.0.5
BcContainerHelper is running as administrator
Host is Microsoft Windows 10 Pro - 21H1
Docker Client Version is 20.10.14
Docker Server Version is 20.10.14
Removing C:\ProgramData\BcContainerHelper\Extensions\mysandbox
Downloading application artifact /sandbox/20.0.37253.39429/ch
Downloading C:\Users\MarcX\AppData\Local\Temp\093359b1-c40c-4fa4-94f4-d90c02d22a60.zip
Unpacking application artifact to tmp folder using 7zip
Fetching all docker images
Fetching all docker volumes
INFO: Windows 10 21H1/21H2 images are not yet available, using 2004 as these are found to work better than 20H2 on 21H1/21H2
Pulling image mcr.microsoft.com/businesscentral:10.0.19041.1415
10.0.19041.1415: Pulling from businesscentral
295f12394c4f: Pulling fs layer
24b98dd3fbc4: Pulling fs layer
569a4c4f5ea8: Pulling fs layer
00dc665dbda6: Pulling fs layer
dd978696e09a: Pulling fs layer
ef2e14c0187f: Pulling fs layer
45eb88e308ea: Pulling fs layer
83c52bf61a50: Pulling fs layer
00dc665dbda6: Waiting
dd978696e09a: Waiting
ef2e14c0187f: Waiting
45eb88e308ea: Waiting
da21618a42d6: Pulling fs layer
aa6427e421e8: Pulling fs layer
33fae7d13c7b: Pulling fs layer
73285156abec: Pulling fs layer
c93b4f5d1462: Pulling fs layer
aa7385607a3f: Pulling fs layer
da21618a42d6: Waiting
2b46fa27c65a: Pulling fs layer
aa6427e421e8: Waiting
33fae7d13c7b: Waiting
aa7385607a3f: Waiting
c93b4f5d1462: Waiting
2b46fa27c65a: Waiting
83c52bf61a50: Waiting
569a4c4f5ea8: Verifying Checksum
569a4c4f5ea8: Download complete
00dc665dbda6: Verifying Checksum
00dc665dbda6: Download complete
dd978696e09a: Verifying Checksum
dd978696e09a: Download complete
ef2e14c0187f: Verifying Checksum
ef2e14c0187f: Download complete
45eb88e308ea: Verifying Checksum
45eb88e308ea: Download complete
83c52bf61a50: Verifying Checksum
83c52bf61a50: Download complete
da21618a42d6: Verifying Checksum
da21618a42d6: Download complete
aa6427e421e8: Verifying Checksum
aa6427e421e8: Download complete
24b98dd3fbc4: Verifying Checksum
24b98dd3fbc4: Download complete
73285156abec: Verifying Checksum
73285156abec: Download complete
c93b4f5d1462: Verifying Checksum
c93b4f5d1462: Download complete
aa7385607a3f: Verifying Checksum
aa7385607a3f: Download complete
2b46fa27c65a: Verifying Checksum
2b46fa27c65a: Download complete
33fae7d13c7b: Verifying Checksum
33fae7d13c7b: Download complete
295f12394c4f: Verifying Checksum
295f12394c4f: Download complete
295f12394c4f: Pull complete
24b98dd3fbc4: Pull complete
569a4c4f5ea8: Pull complete
00dc665dbda6: Pull complete
dd978696e09a: Pull complete
ef2e14c0187f: Pull complete
45eb88e308ea: Pull complete
83c52bf61a50: Pull complete
da21618a42d6: Pull complete
aa6427e421e8: Pull complete
33fae7d13c7b: Pull complete
73285156abec: Pull complete
c93b4f5d1462: Pull complete
aa7385607a3f: Pull complete
2b46fa27c65a: Pull complete
Digest: sha256:5081759663a6e853232a99c2f7289d19e384ac148abe637e147d58739fc281e1
Status: Downloaded newer image for mcr.microsoft.com/businesscentral:10.0.19041.1415
mcr.microsoft.com/businesscentral:10.0.19041.1415
Using image mcr.microsoft.com/businesscentral:10.0.19041.1415
Creating Container mysandbox
Style: sandbox
Multitenant: Yes
Version: 20.0.37253.39429
Platform: 20.0.37114.39316
Generic Tag: 1.0.2.1
Container OS Version: 10.0.19041.1415 (2004)
Host OS Version: 10.0.19043.1645 (21H1)
WARNING: Host OS is 21H1 and Container OS is 2004, defaulting to process isolation. If you experience problems, add -isolation hyperv.
Using process isolation
Using locale de-CH
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Files in C:\ProgramData\BcContainerHelper\Extensions\mysandbox\my:
- AdditionalOutput.ps1
- MainLoop.ps1
- SetupVariables.ps1
- updatecontainerhosts.ps1
Creating container mysandbox from image mcr.microsoft.com/businesscentral:10.0.19041.1415
9268ee536cbb0ef4c04eb2bb0796916cda6903268cd2e59571fa89f539d16ee1
New-BCContainer Telemetry Correlation Id: ebb275a9-0657-46d0-b58b-2e96059298c5
DockerDo : docker: Error response from daemon: hcsshim::PrepareLayer - failed failed in Win32: Incorrect function. (0x1).
ExitCode: 125
Commandline: docker run --volume "c:\bcartifacts.cache:c:\dl" --label nav= --env isBcSandbox=Y --label version=20.0.37253.39429 --label platform=20.0.37114.39316 --label country=CH --env 
artifactUrl=https://bcartifacts.azureedge.net/sandbox/20.0.37253.39429/ch --env multitenant=Y --env licenseFile="" --name mysandbox --hostname mysandbox --env auth=Windows --env username="MarcX" --env ExitOnError=N --env 
locale=de-CH --env databaseServer="" --env databaseInstance="" --volume "C:\ProgramData\BcContainerHelper:C:\ProgramData\BcContainerHelper" --volume "C:\ProgramData\BcContainerHelper\Extensions\mysandbox\my:C:\Run\my" 
--isolation process --restart unless-stopped --env filesOnly=False --env enableApiServices=Y --env useSSL=N --env securePassword=xxxxx --env passwordKeyFile="c:\run\my\aes.key" --env removePasswordKeyFile=Y --env accept_eula=Y --env accept_outdated=Y --detach mcr.microsoft.com/businesscentral:10.0.19041.1415
In C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.5\ContainerHandling\New-NavContainer.ps1:1828 Zeichen:15
+ ...       if (!(DockerDo -accept_eula -accept_outdated:$accept_outdated - ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,DockerDo

I removed all the images and reinstalled docker, the result is the same (happends all the time).

What is wrong?
BR
Marc

**Comment by freddydk** [comment]
First thing I would try would be -isolation hyperv

**Comment by MarcFB** [comment]
unfortunately it gives the same error:

PS C:\WINDOWS\system32> $artifactUrl = Get-BcArtifactUrl -type sandbox -country ch -select Latest
New-BCContainer -isolation hyperv -accept_eula -containerName mysandbox -artifactUrl $artifactUrl
BcContainerHelper is version 3.0.5
BcContainerHelper is running as administrator
Host is Microsoft Windows 10 Pro - 21H1
Docker Client Version is 20.10.14
Docker Server Version is 20.10.14
Removing C:\ProgramData\BcContainerHelper\Extensions\mysandbox
Downloading application artifact /sandbox/20.0.37253.39459/ch
Downloading C:\Users\MarcX\AppData\Local\Temp\e4ae8852-a2c7-4428-9a9b-e14a35f4c740.zip
Unpacking application artifact to tmp folder using 7zip
Downloading platform artifact /sandbox/20.0.37253.39459/platform
Downloading C:\Users\MarcX\AppData\Local\Temp\d38139ff-1306-4466-8210-a20fedc010a3.zip
Unpacking platform artifact to tmp folder using 7zip
Downloading Prerequisite Components
Downloading c:\bcartifacts.cache\sandbox\20.0.37253.39459\platform\Prerequisite Components\Open XML SDK 2.5 for Microsoft Office\OpenXMLSDKv25.msi
Downloading c:\bcartifacts.cache\sandbox\20.0.37253.39459\platform\Prerequisite Components\IIS URL Rewrite Module\rewrite_2.0_rtw_x64.msi
Downloading c:\bcartifacts.cache\sandbox\20.0.37253.39459\platform\Prerequisite Components\DotNetCore\DotNetCore.1.0.4_1.1.1-WindowsHosting.exe
Fetching all docker images
Fetching all docker volumes
INFO: Windows 10 21H1/21H2 images are not yet available, using 2004 as these are found to work better than 20H2 on 21H1/21H2
Using image mcr.microsoft.com/businesscentral:10.0.19041.1415
Creating Container mysandbox
Style: sandbox
Multitenant: Yes
Version: 20.0.37253.39459
Platform: 20.0.37114.39436
Generic Tag: 1.0.2.1
Container OS Version: 10.0.19041.1415 (2004)
Host OS Version: 10.0.19043.1645 (21H1)
Using hyperv isolation
Using locale de-CH
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Files in C:\ProgramData\BcContainerHelper\Extensions\mysandbox\my:
- AdditionalOutput.ps1
- MainLoop.ps1
- SetupVariables.ps1
- updatecontainerhosts.ps1
Creating container mysandbox from image mcr.microsoft.com/businesscentral:10.0.19041.1415
dbeb3d903af83bcb2bc61fc6a82e92bc37d513d0752ae3ced8141c3314fd57d5
New-BCContainer Telemetry Correlation Id: 0ecfa428-7099-48dd-bd4e-f7413cbc714d
DockerDo : docker: Error response from daemon: hcsshim::CreateComputeSystem dbeb3d903af83bcb2bc61fc6a82e92bc37d513d0752ae3ced8141c3314fd57d5: The requested resource is in use.
ExitCode: 125
Commandline: docker run --volume "c:\bcartifacts.cache:c:\dl" --label nav= --env isBcSandbox=Y --label version=20.0.37253.39459 --label platform=20.0.37114.39436 --label country=CH --env 
artifactUrl=https://bcartifacts.azureedge.net/sandbox/20.0.37253.39459/ch --env multitenant=Y --env licenseFile="" --name mysandbox --hostname mysandbox --env auth=Windows --env username="MarcX" --env ExitOnError=N --env 
locale=de-CH --env databaseServer="" --env databaseInstance="" --volume "C:\ProgramData\BcContainerHelper:C:\ProgramData\BcContainerHelper" --volume "C:\ProgramData\BcContainerHelper\Extensions\mysandbox\my:C:\Run\my" 
--isolation hyperv --restart unless-stopped --env filesOnly=False --memory 8G --env enableApiServices=Y --env useSSL=N --env securePassword=xxx --env passwordKeyFile="c:\run\my\aes.key" --env removePasswordKeyFile=Y --env accept_eula=Y --env accept_outdated=Y --detach mcr.microsoft.com/businesscentral:10.0.19041.1415
In C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.5\ContainerHandling\New-NavContainer.ps1:1828 Zeichen:15
+ ...       if (!(DockerDo -accept_eula -accept_outdated:$accept_outdated - ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,DockerDo

**Comment by freddydk** [comment]
Then I think you should try this:
https://freddysblog.com/2018/12/11/clean-up-after-yourself-docker-your-mom-isnt-here/
Which cleans up the docker filesystem (if there are cross linked files or like)

**Comment by MarcFB** [comment]
I did the "Reset To Factory defaults" you described and reinstalled everything. The error is still there :-(

As mentioned in the error description, I also tried to reinstall everything ... same problem.

**Comment by freddydk** [comment]
The blog post mentions running the docker zap script - did you do that?

**Comment by MarcFB** [comment]
No, it was not clear to me that this is still necessary. I'm running it now...

**Comment by MarcFB** [comment]
Next error:

PS C:\WINDOWS\system32> C:\Users\MarcX\Unbenannt5.ps1
Checking Docker Service Settings...
This function will remove all containers, remove all images and clear the folder C:\ProgramData\Docker
The function will also clear the contents of C:\ProgramData\BcContainerHelper.
Are you absolutely sure you want to do this? (This cannot be undone)
Type Yes to continue:Yes
Running Docker System Prune
Total reclaimed space: 0B
Removing all containers (forced)
Stopping Docker Service
Downloading Docker-Ci-Zap
Running Docker-Ci-Zap on C:\ProgramData\Docker
Note: If this fails, please restart your computer and run this script again
INFO: Zapped successfully
Removing Docker-Ci-Zap
Starting Docker Service
Start-Service : Fehler beim Starten des Diensts "Docker Engine (docker)".
In C:\Users\MarcX\Unbenannt5.ps1:66 Zeichen:1
+ Start-Service docker
+ ~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OpenError: (System.ServiceProcess.ServiceController:ServiceController) [Start-Service], ServiceCommandException
    + FullyQualifiedErrorId : StartServiceFailed,Microsoft.PowerShell.Commands.StartServiceCommand

And the Eventlog shows:
fatal: unable to configure the Docker daemon with file C:\ProgramData\Docker\config\daemon.json: open C:\ProgramData\Docker\config\daemon.json: The system cannot find the path specified.

does the skrip remove something it needs to start the process?

**Comment by freddydk** [comment]
It should not - does a restart change anything?`
else you will have to uninstall and reinstall again.

**Comment by MarcFB** [comment]
After a restart, the system seems to recreate the c:\programData\Docker folder and the subfolder/files. I reinstalled bccontainerhelper and the tried to reinstall the image:same result (same error message)

**Comment by freddydk** [comment]
same error message - like in hcsshim::PrepareLayer - failed failed in Win32: Incorrect function. (0x1)
or hcsshim::CreateComputeSystem - The requested resource is in use.

The first one is typically due to incompatible OS between container and host.
The second one have been seen quite a lot and typically is due to lack of memory or paging file too small (cleanup some times fixes this)

If you cannot make it work with this - my recommendation would be to upgrade to Windows 11

**Comment by MarcFB** [comment]
I managed to get around that error message. I terminated all services that are not essential (eg Chrome, Teams, ...) and then the process went on until it stopped with an error message when trying to start the container.
Running this image on a Windows 10 installation seems to be very nitpicky! I'll try whether I can use a azure container (server image).
Thanks for your help
Br
Marc

**Comment by MarcFB** [comment]
Sorry to bother you again but I also failed to make it work on azure :-(
I tried exactly the steps as described here:
https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-get-started-container-sandbox

All I get ist the following error message from azure:
"...
The template deployment failed with error: 'The resource with id: '/subscriptions/d98a32aa-2c89-461e-a4c8-53bd2b5e84b6/resourceGroups/myNAVRG/providers/Microsoft.Compute/virtualMachines/test' failed validation with message: 'The requested size for resource '/subscriptions/d98a32aa-2c89-461e-a4c8-53bd2b5e84b6/resourceGroups/myNAVRG/providers/Microsoft.Compute/virtualMachines/test' is currently not available in location 'switzerlandnorth' zones '' for subscription 'd98a32aa-2c89-461e-a4c8-53bd2b5e84b6'. Please try another size or deploy to a different location or zones. See https://aka.ms/azureskunotavailable for details.'.'. (Code: InvalidTemplateDeployment)
.."

It's really frustrating that nothing works  - not even the basic examples that MS provides :-(
Did you ever manage to bring one of those sandbox examples up and running?

**Comment by MarcFB** [comment]
To give you some more details:
- I open the "Container Sandbox Environment" page in BC (cloud version) and click on "Host in Azure".
- This opens a new page --> https://portal.azure.com/#create/Microsoft.Template
-  Here are my settings
![image](https://user-images.githubusercontent.com/47138723/165323022-d321b9b1-1a3c-421b-a7df-c6a6af40ff8b.png)
- now i cklick on check and Proceed and I get 

{"code":"InvalidTemplateDeployment","message":"The template deployment failed with error: 'The resource with id: '/subscriptions/d98a32aa-2c89-461e-a4c8-53bd2b5e84b6/resourceGroups/myNAVRG/providers/Microsoft.Compute/virtualMachines/test' failed validation with message: 'The requested size for resource '/subscriptions/d98a32aa-2c89-461e-a4c8-53bd2b5e84b6/resourceGroups/myNAVRG/providers/Microsoft.Compute/virtualMachines/test' is currently not available in location 'switzerlandnorth' zones '' for subscription 'd98a32aa-2c89-461e-a4c8-53bd2b5e84b6'. Please try another size or deploy to a different location or zones. See https://aka.ms/azureskunotavailable for details.'.'."}

I tried different regions, the result is always the same

**Comment by freddydk** [comment]
The error states that the requested size is not available in switzerlandnorth.
You can try a different datacenter or see which sizes are available in the URL provided in the error.
I typically use North Europe or West Europe and have never seen these problems.

**Comment by freddydk** [comment]
Have you tried creating just a simple VM to see whether your subscription allows you to add a VM?

**Comment by MarcFB** [comment]
I tried to create a VM in the azure web interface and it seems that I can do it in "westeurope".

Using "westeurope" in the BC skript leads to the same error:

{"code":"InvalidTemplateDeployment","message":"The template deployment failed with error: 'The resource with id: '/subscriptions/d98a32aa-2c89-461e-a4c8-53bd2b5e84b6/resourceGroups/BC_Competec/providers/Microsoft.Compute/virtualMachines/bc_competec' failed validation with message: 'The requested size for resource '/subscriptions/d98a32aa-2c89-461e-a4c8-53bd2b5e84b6/resourceGroups/BC_Competec/providers/Microsoft.Compute/virtualMachines/bc_competec' is currently not available in location 'westeurope' zones '' for subscription 'd98a32aa-2c89-461e-a4c8-53bd2b5e84b6'. Please try another size or deploy to a different location or zones. See https://aka.ms/azureskunotavailable for details.'.'."}

I tried to ge a list of the valid VMs with 
az vm list-skus --location westeurope --output table

and there are quite some available VMs. But there is no match between the available VMs in azure and the VMs I can choose in the skript:

![image](https://user-images.githubusercontent.com/47138723/165815434-272532ce-a68a-4d39-8c8b-0cde53c4ad09.png)

**Comment by MarcFB** [comment]
Setting up a VM with the web interface works
![image](https://user-images.githubusercontent.com/47138723/165816552-6cdc477e-d8f3-40d9-bdde-ea74adb8146d.png)

**Comment by freddydk** [comment]
Very strange - it seems like you are missing a number of images????
When I run your az account - it looks like this:
![image](https://user-images.githubusercontent.com/10775043/165897059-86c785d4-c804-44e4-b030-1b1c14118dec.png)

Is there anything special with that subscription?

I will see if I can make the list in the ARM template more dynamic - or in other ways allow people to specify other sizes.

**Comment by freddydk** [comment]
I have updated the available sizes in the ARM templates (https://aka.ms/getbc).
In your list - you only had the DC VMs (confidential = more expensive) or the AMD based VMs v5 (Dxxas_v5)
The DC VMs won't be in the list - I have added _v4 and _v5 (including now AMD based VMs as well)

**Comment by Roglar01** [comment]
Can confirm having the same issue, but the using
`-myscripts @("https://raw.githubusercontent.com/microsoft/nav-docker/master/override/issue2434/start.ps1","https://raw.githubusercontent.com/microsoft/nav-docker/master/override/issue2434/navinstall.ps1")` as a workaround is a success.
I'm using Windows 10 Enterprise -21H2

**Comment by freddydk** [comment]
Did you manage to make this work?

**Comment by MarcFB** [comment]
Hi Freddy
Not yet! I had a week holidays 😊

Right now I try to bring an docker image on a Windows Server 19 up and running. I managed to install the image but I did not manage to connect Visual Studio Code
There are always  billions of questions and even more errors and problems ☹

For example:
Visual Studio Code says
[2022-05-11 13:39:29.92] Using reference symbols cache path: c:\Users\mkt1\Documents\AL\BCTest\./.alpackages
[2022-05-11 13:39:29.92] Sending request to http://bc-mk:8080/BC/dev/metadata?tenant=default
[2022-05-11 13:39:30.58] The request for path /BC/dev/metadata?tenant=default failed with code NotFound. Reason: Not Found
[2022-05-11 13:39:30.58] Sending request to http://bc-mk:8080/BC/dev/webendpoint
[2022-05-11 13:39:30.58] The request for path /BC/dev/webendpoint failed with code NotFound. Reason: Not Found
[2022-05-11 13:39:30.58] Sending request to http://bc-mk:8080/BC/dev/metadata?tenant=default
[2022-05-11 13:39:30.58] The request for path /BC/dev/metadata?tenant=default failed with code NotFound. Reason: Not Found
[2022-05-11 13:39:30.58] Sending request to http://bc-mk:8080/BC/dev/webendpoint
[2022-05-11 13:39:30.58] The request for path /BC/dev/webendpoint failed with code NotFound. Reason: Not Found

When I try to download symbols. What’s wrong?
In you blog https://freddysblog.com/2019/08/20/business-central-on-docker-for-non-experts/
You describe that with the option -includeAL is helpful when I want to do customizations. Guess what happens when I add that switch to New-BCContainer?

-includeAL : The term '-includeAL' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path
was included, verify that the path is correct and try again.
At line:42 char:3
+         -includeAL
+         ~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (-includeAL:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

I also have no idea how I can access the BC administration (e.g for switching to SSL)

It’s all broken, untested, outdated, …



Von: Freddy Kristiansen ***@***.***>
Gesendet: Mittwoch, 11. Mai 2022 06:40
An: microsoft/navcontainerhelper ***@***.***>
Cc: Marc Knaff ***@***.***>; Author ***@***.***>
Betreff: Re: [microsoft/navcontainerhelper] Error Creating a container with BCContainerHelper -- New-NavContainer.ps1:1828 Character:15 (Issue #2419)


Did you manage to make this work?

**Comment by freddydk** [comment]
it looks like you added a new line with -includeAL without adding a ` character to end the preceding line - that causes powershell to do what you describe.
-includeAL is only really used for code customized base apps, I don't think you are going there.

I know it is tedious, but you should in every comment include the full script and the full output of that, that makes it much easier to help.

Also - in VS Code - what is the output of launch.json ?
Looks like you have set the port to 8080 - that is likely not correct.

**Comment by MarcFB** [comment]
You are right, I forgot the ‘ in the script ☹
Now I tried without port and got an authentication error (similar to https://github.com/microsoft/AL/issues/6273)

But let’s start with the dodumentation:

I use the following script:

$accept_eula = $true
$containername = 'MKBC2022'
$licenseFile = ''
$auth = 'Windows'
$usessl = $false
$updateHosts = $true
$assignPremiumPlan = $true
$shortcuts = 'Desktop'

$artifactUrl = Get-BcArtifactUrl -country 'CH' -select 'Latest'

if ($auth -eq 'Windows') {
    $credential = get-credential -UserName $env:USERNAME -Message 'Using Windows Authentication. Please enter your Windows credentials.'
}
elseif ($auth -eq 'UserPassword') {
    $credential = get-credential -Message 'Using Username/Password Authentication. Please enter credentials.'
}
else {
    throw '$auth needs to be either Windows or UserPassword'
}

if (!$accept_eula) {
    throw 'You have to accept the eula in order to create a Container.'
}
if (!($containername)) {
    throw 'You have to specify a container name in order to create a container.'
}
if (!($artifactUrl)) {
    throw 'No matching build was found for your version of Business Central'
}
New-BcContainer -accept_eula:$accept_eula `
                -containername $containername `
                -artifactUrl $artifactUrl `
                -auth $auth `
                -Credential $credential `
                -usessl:$usessl `
                -updateHosts:$updateHosts `
                -licenseFile $licenseFile `
                -assignPremiumPlan:$assignPremiumPlan `
                -shortcuts $shortcuts

if ($licenseFile) {
    Setup-BcContainerTestUsers -containerName $containername -password $credential.Password
}

This is the output

BcContainerHelper is version 3.0.7
BcContainerHelper is running as administrator
Hyper-V is Disabled
Host is Microsoft Windows Server 2019 Standard - ltsc2019
Docker Client Version is 20.10.9
Docker Server Version is 20.10.9
Downloading application artifact /sandbox/20.1.39764.40432/ch
Downloading C:\Users\mkt1\AppData\Local\Temp\2\47149270-f527-4486-ae38-f90f8a6df826.zip
Unpacking application artifact to tmp folder using Expand-Archive
Downloading platform artifact /sandbox/20.1.39764.40432/platform
Downloading C:\Users\mkt1\AppData\Local\Temp\2\ec59ae5d-7414-4d6a-a1f2-5b4f5301efb9.zip
Unpacking platform artifact to tmp folder using Expand-Archive
Downloading Prerequisite Components
Downloading c:\bcartifacts.cache\sandbox\20.1.39764.40432\platform\Prerequisite Components\Open XML SDK 2.5 for Microsoft Office\OpenXMLSDKv25.msi
Downloading c:\bcartifacts.cache\sandbox\20.1.39764.40432\platform\Prerequisite Components\IIS URL Rewrite Module\rewrite_2.0_rtw_x64.msi
Downloading c:\bcartifacts.cache\sandbox\20.1.39764.40432\platform\Prerequisite Components\DotNetCore\DotNetCore.1.0.4_1.1.1-WindowsHosting.exe
Fetching all docker images
Fetching all docker volumes
Using image mcr.microsoft.com/businesscentral:10.0.17763.973
Creating Container MKBC2022
Style: sandbox
Multitenant: Yes
Version: 20.1.39764.40432
Platform: 20.0.39668.40349
Generic Tag: 1.0.2.1
Container OS Version: 10.0.17763.973 (ltsc2019)
Host OS Version: 10.0.17763.973 (ltsc2019)
Using process isolation
Using locale de-CH
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Files in C:\ProgramData\BcContainerHelper\Extensions\MKBC2022\my:
- AdditionalOutput.ps1
- MainLoop.ps1
- SetupNavUsers.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container MKBC2022 from image mcr.microsoft.com/businesscentral:10.0.17763.973
17ca69b129526c7c8ced686c2ba4e5198f9c63c58f48678a6c63f2b967c633c2
Waiting for container MKBC2022 to be ready
Using artifactUrl https://bcartifacts.azureedge.net/sandbox/20.1.39764.40432/ch
Using installer from C:\Run\150-new
Installing Business Central
Installing from artifacts
Starting Local SQL Server
Starting Internet Information Server
Copying Service Tier Files
c:\dl\sandbox\20.1.39764.40432\platform\ServiceTier\Program Files
c:\dl\sandbox\20.1.39764.40432\platform\ServiceTier\System64Folder
Copying PowerShell Scripts
c:\dl\sandbox\20.1.39764.40432\platform\WindowsPowerShellScripts\Cloud\NAVAdministration
c:\dl\sandbox\20.1.39764.40432\platform\WindowsPowerShellScripts\WebSearch
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Determining Database Collation from c:\dl\sandbox\20.1.39764.40432\ch\BusinessCentral-CH.bak
Restoring CRONUS Demo Database
Setting CompatibilityLevel for tenant on localhost\SQLEXPRESS
Exporting Application to CRONUS
Removing Application from tenant
Modifying Business Central Service Tier Config File for Docker
Creating Business Central Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Copying Web Client Files
c:\dl\sandbox\20.1.39764.40432\platform\WebClient\Microsoft Dynamics NAV
Copying Client Files
c:\dl\sandbox\20.1.39764.40432\platform\LegacyDlls\program files\Microsoft Dynamics NAV
c:\dl\sandbox\20.1.39764.40432\platform\LegacyDlls\program files\Microsoft Dynamics NAV
c:\dl\sandbox\20.1.39764.40432\platform\LegacyDlls\systemFolder
Copying ModernDev Files
c:\dl\sandbox\20.1.39764.40432\platform
c:\dl\sandbox\20.1.39764.40432\platform\ModernDev\program files\Microsoft Dynamics NAV
Copying additional files
Copying ConfigurationPackages
C:\dl\sandbox\20.1.39764.40432\ch\ConfigurationPackages
Copying Test Assemblies
C:\dl\sandbox\20.1.39764.40432\platform\Test Assemblies
Copying Extensions
C:\dl\sandbox\20.1.39764.40432\ch\Extensions
Copying Applications
C:\dl\sandbox\20.1.39764.40432\platform\Applications
Copying Applications.CH
C:\dl\sandbox\20.1.39764.40432\ch\Applications.CH
Starting Business Central Service Tier
Importing license file
Copying Database on localhost\SQLEXPRESS from tenant to default
Taking database tenant offline
Copying database files
Attaching files as new Database default
Putting database tenant back online
Mounting tenant database
Mounting Database for default on server localhost\SQLEXPRESS with AllowAppDatabaseWrite = False
Sync'ing Tenant
Tenant is Operational
Stopping Business Central Service Tier
Installation took 152 seconds
Installation complete
Initializing...
Setting host.containerhelper.internal to 172.24.224.1 in container hosts file
Starting Container
Hostname is MKBC2022
PublicDnsName is MKBC2022
Using Windows Authentication
Modifying Service Tier Config File with Instance Specific Settings
Starting Service Tier
Registering event sources
Creating DotNetCore Web Server Instance
Enabling Financials User Experience
Dismounting Tenant
Mounting Tenant
Mounting Database for default on server localhost\SQLEXPRESS with AllowAppDatabaseWrite = False
Sync'ing Tenant
Tenant is Operational
Creating http download site
Creating Windows user mkt1
Setting SA Password and enabling SA
Creating SUPER user
Assign Premium plan for MKBC2022\MKT1
Container IP Address: 172.24.231.159
Container Hostname  : MKBC2022
Container Dns Name  : MKBC2022
Web Client          : http://MKBC2022/BC/?tenant=default
Dev. Server         : http://MKBC2022
Dev. ServerInstance : BC
Dev. Server Tenant  : default
Setting MKBC2022 to 172.24.231.159 in host hosts file
Setting MKBC2022-default to 172.24.231.159 in host hosts file
Setting MKBC2022-default to 172.24.231.159 in container hosts file

Files:
http://MKBC2022:8080/ALLanguage.vsix

WARNING: You are running a container which is 535 days old.
Microsoft recommends that you always run the latest version of our containers.

Container Total Physical Memory is 16.0Gb
Container Free Physical Memory is 8.9Gb

Initialization took 39 seconds
Ready for connections!
Reading CustomSettings.config from MKBC2022
Creating Desktop Shortcuts for MKBC2022
Container MKBC2022 successfully created

Use:
Get-BcContainerEventLog -containerName MKBC2022 to retrieve a snapshot of the event log from the container
Get-BcContainerDebugInfo -containerName MKBC2022 to get debug information about the container
Enter-BcContainer -containerName MKBC2022 to open a PowerShell prompt inside the container
Remove-BcContainer -containerName MKBC2022 to remove the container again
docker logs MKBC2022 to retrieve information about URL's again

My Visual Studio Code looks like

***@***.***

And you see the output in the “Output” section
I can access BC with the link
http://mkbc2022/BC/?tenant=default

And this is how the user looks:
***@***.***


Do you have any idea, what’s the problem now?
BR
Marc


Von: Freddy Kristiansen ***@***.***>
Gesendet: Mittwoch, 11. Mai 2022 14:49
An: microsoft/navcontainerhelper ***@***.***>
Cc: Marc Knaff ***@***.***>; Author ***@***.***>
Betreff: Re: [microsoft/navcontainerhelper] Error Creating a container with BCContainerHelper -- New-NavContainer.ps1:1828 Character:15 (Issue #2419)


it looks like you added a new line with -includeAL without adding a ` character to end the preceding line - that causes powershell to do what you describe.
-includeAL is only really used for code customized base apps, I don't think you are going there.

I know it is tedious, but you should in every comment include the full script and the full output of that, that makes it much easier to help.

Also - in VS Code - what is the output of launch.json ?
Looks like you have set the port to 8080 - that is likely not correct.

**Comment by MarcFB** [comment]
No problem :-)

Visual Studio Code:
![image](https://user-images.githubusercontent.com/47138723/167874936-bbeb62fc-16b7-43b1-9ba0-f333e425cca0.png)

User Card in BC
![image](https://user-images.githubusercontent.com/47138723/167875000-d9b7bc1d-ecda-4e93-9686-1a7d1e1f2e0a.png)

**Comment by freddydk** [comment]
You container is setup with Windows authentication.
in launch.json you have specified UserPassword authentication.

Either recreate your container with userpassword - or change to Windows in launch.json
Note for Windows Auth to work - the container needs to be started with your Windows credentials
Lastly windows Auth. is known to cause problems - i always recommend userpassword.

**Comment by freddydk** [comment]
Did you manage to make this work?

**Comment by MarcFB** [comment]
Hi Freddy
Yes, I was able to make it work. The problem was - like you supposed - the different authentication methods. Since I used the script which is provided on the "sandox environment" page in BC, I did not pay attention to this.

I also was able to set up a Azure Virtual Machine with the new scrip you provided (let's say the new bvirtual machines you added). I found a match in region europewest.

So finally I have my sandbox :-)

Thank you for your help
BR
Marc
